lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java

```diff
@@ -2262,4 +2351,117 @@ public void onError(ErrorInfo reason) {
 		}
 	}
 
+	/**
+	 * Test if presence sync works as it should
+	 * Tests RTP18a, RTP18b, RTP2f
+	 */
+	@Test
+	public void presence_sync() {
+		AblyRealtime ably = null;
+		try {
+			ClientOptions opts = createOptions(testVars.keys[0].keyStr);
+			ably = new AblyRealtime(opts);
+
+			final String channelName = "presence_sync_test" + testParams.name;
+
+			final Channel channel = ably.channels.get(channelName);
+			channel.attach();
+			ChannelWaiter channelWaiter = new ChannelWaiter(channel);
+			channelWaiter.waitFor(ChannelState.attached);
+
+			final PresenceMessage[] testPresence1 = new PresenceMessage[] {
+					/* Will be discarded because we'll start new sync with different channelSerial */
+					new PresenceMessage() {{
+						clientId = "1";
+						action = Action.enter;
+						connectionId = "1";
+						id = "1:0";
+					}}
+			};
+
+			final PresenceMessage[] testPresence2 = new PresenceMessage[] {
+					new PresenceMessage() {{
+						clientId = "2";
+						action = Action.enter;
+						connectionId = "2";
+						id = "2:1:0";
+					}},
+					/* Enter presence message here is newer than leave in the subsequent message */
+					new PresenceMessage() {{
+						clientId = "3";
+						action = Action.enter;
+						connectionId = "3";
+						id = "3:1:0";
+					}}
+			};
+
+			final PresenceMessage[] testPresence3 = new PresenceMessage[] {
+					new PresenceMessage() {{
+						clientId = "3";
+						action = Action.leave;
+						connectionId = "3";
+						id = "3:0:0";
+					}},
+					new PresenceMessage() {{
+						clientId = "4";
+						action = Action.enter;
+						connectionId = "4";
+						id = "4:1:1";
+					}},
+					new PresenceMessage() {{
+						clientId = "4";
+						action = Action.leave;
+						connectionId = "4";
+						id = "4:2:2";
+					}}
+			};
+
+			channel.presence.subscribe(Action.leave, new Presence.PresenceListener() {
+				@Override
+				public void onPresenceMessage(PresenceMessage message) {
+					try {
+						if (message.clientId.equals("4") && message.action == Action.leave)
+							assertEquals("Verify LEAVE message is recorded as ABSENT during sync",
```

Comment:
How do you even know this assertion ran?  This test could complete without this having run no?